test/e2e/framework/nodes.go

```diff
@@ -0,0 +1,42 @@
+package framework
+
+import (
+	"strconv"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (f *Framework) FindNodes(cluster ClusterIndex) (gateway *v1.Node, nonGateway *v1.Node) {
+	nodes := AwaitUntil("list nodes", func() (interface{}, error) {
+		// Ignore the control plane node labeled as master as it doesn't allow scheduling of pods
+		return f.ClusterClients[cluster].CoreV1().Nodes().List(metav1.ListOptions{
+			LabelSelector: "!node-role.kubernetes.io/master",
+		})
+	}, NoopCheckResult).(*v1.NodeList)
+
+	var gatewayName, nonGatewayName string
+	for i := range nodes.Items {
+		if nodes.Items[i].Labels[GatewayLabel] == "true" {
+			gateway = &nodes.Items[i]
+			gatewayName = gateway.Name
+		} else {
+			nonGateway = &nodes.Items[i]
+			nonGatewayName = nonGateway.Name
+		}
+	}
+
+	Logf("Found gateway node: %s, non-gateway node: %s", gatewayName, nonGatewayName)
+	return
+}
+
+func (f *Framework) SetGatewayLabelOnNode(cluster ClusterIndex, nodeName string, isGateway bool) {
```

Comment:
How about adding a "UnsetGatewayLabelOnNode" for remove cluster test use case?

Reply:
Using the same function with isGateway "true" or "false" is enough. Renaming the function to "ToggleGatewayLabelOnNode" would be more appropriate.

This is not something that needs urgent change.

Rest of the patch looks good.

Reply:
Set is better here, toggle implies that you don’t pass true/false and It will switch the existing state.